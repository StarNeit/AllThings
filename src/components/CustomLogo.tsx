import React, { useEffect, useState } from 'react'
import { Theme, View } from '@allthings/elements'

const SIZE = 32.5 // same as the Element's Icon.

interface IProps {
  url: string
  color: string
}

const CustomLogo = ({ url, color }: IProps) => {
  // Use a fallback url (external content icon). This should actually never be
  // used since we are the ones responsible for uploading the custom SVG icons!
  const fallbackUrl =
    'https://static.allthings.me/react-icons/production/logoutFilled.svg'
  const [unsafeHTMLContent, setUnsafeHTMLContent] = useState('')

  // TODO: Use Suspense when stable and available.
  // See: https://stackoverflow.com/a/53572588
  const fetchCustomLogo = async (customLogoUrl: string) => {
    const response = await fetch(customLogoUrl)

    if (response.status === 200) {
      setUnsafeHTMLContent(await response.text())
    } else {
      const fallbackResponse = await fetch(fallbackUrl)

      // We assume that the fallback is always available!
      setUnsafeHTMLContent(await fallbackResponse.text())
    }
  }

  useEffect(() => {
    fetchCustomLogo(url)
  }, [url])

  return (
    <Theme>
      {({ colorize }: { colorize: (color: string) => string }) => (
        <View
          alignH="center"
          alignV="center"
          dangerouslySetInnerHTML={{ __html: unsafeHTMLContent }}
          style={{
            fill: colorize(color),
            height: SIZE,
            width: SIZE,
          }}
        />
      )}
    </Theme>
  )
}

export default CustomLogo
