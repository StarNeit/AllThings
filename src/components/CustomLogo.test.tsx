import React from 'react'
import CustomLogo from './CustomLogo'
import { cleanup, render, wait } from '@testing-library/react'

const COLOR = '#fff'
const URL = 'https://app.dev.allthings.me/test-icon.svg'
const FALLBACK_URL =
  'https://static.allthings.me/react-icons/production/logoutFilled.svg'

// Sadly Enzyme doesn't support hooks yet!
// https://github.com/airbnb/enzyme/issues/2011
// And react-testing-library has a verbosity issue
// https://github.com/kentcdodds/react-testing-library/issues/281

afterEach(cleanup)

describe('Check the CustomLogo component', () => {
  beforeEach(() => {
    global.fetch.resetMocks()
  })
  it('should fetch and retry again with a fallback', async () => {
    global.fetch
      .mockResponseOnce(JSON.stringify({}), {
        status: 404,
      })
      .mockResponseOnce(JSON.stringify({ status: 200 }))
    const { container } = render(<CustomLogo color={COLOR} url={URL} />)
    expect(container).toMatchSnapshot()
    expect(global.fetch.mock.calls.length).toEqual(1)
    expect(global.fetch.mock.calls[0][0]).toEqual(URL)
    await wait(() => {
      expect(global.fetch.mock.calls.length).toEqual(2)
      expect(global.fetch.mock.calls[1][0]).toEqual(FALLBACK_URL)
    })
  })
})
