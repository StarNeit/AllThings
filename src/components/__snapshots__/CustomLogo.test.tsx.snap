// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the CustomLogo component should fetch and retry again with a fallback 1`] = `
.css-0,
[data-css-0] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

<div>
  <div
    data-css-0=""
    style="fill: #fff; height: 32.5px; width: 32.5px;"
  />
</div>
`;
